cdap-api/src/main/java/co/cask/cdap/api/metadata/MetadataEntity.java

```diff
@@ -74,88 +83,160 @@
-    return new MetadataEntity(DATASET, datasetName);
+  private MetadataEntity(Map<String, String> details, String type) {
+    this.details = details;
+    this.type = type;
   }
 
   /**
-   * Creates a {@link MetadataEntity} representing the given datasetName in the specified namespace.
-   *
-   * @param namespace the name of the namespace
-   * @param datasetName the name of the dataset
-   * @return {@link MetadataEntity} representing the dataset name
+   * Builder for {@link MetadataEntity}
    */
-  public static MetadataEntity ofDataset(String namespace, String datasetName) {
-    return MetadataEntity.ofNamespace(namespace).appendAsType(MetadataEntity.DATASET, datasetName);
+  public static class Builder {
+    private final LinkedHashMap<String, String> parts;
+    private String type;
+
+    Builder() {
+      parts = new LinkedHashMap<>();
+    }
+
+    private Builder(MetadataEntity metadataEntity) {
+      this.parts = new LinkedHashMap<>(metadataEntity.details);
+      this.type = metadataEntity.type;
+    }
+
+    /**
+     * Put the given key with the given value in {@link MetadataEntity.Builder}.
+     * The returned {@link MetadataEntity.Builder} is of the same type. If the type needs to
+     * changed during append then {@link #appendAsType(String, String)}  should be used.
+     *
+     * @param key the key to be added
+     * @param value the value to be added
+     * @return a new {@link MetadataEntity.Builder} which is of same type of this {@link MetadataEntity.Builder}
+     * but consists of the given key and value in addition
+     */
+    public Builder append(String key, String value) {
+      parts.put(key.toLowerCase(), value);
+      return this;
+    }
+
+    /**
+     * Put the given key with the given value in {@link MetadataEntity.Builder}.
+     * The returned {@link MetadataEntity} type is set to the given key. If an append is required without the
+     * type change then {@link #append(String, String)} should be used.
+     *
+     * @param key the key to be added
+     * @param value the value to be added
+     * @return a new {@link MetadataEntity.Builder} whose type is set to the given key and consists of the given
+     * key and value in addition
+     */
+    public Builder appendAsType(String key, String value) {
+      parts.put(key.toLowerCase(), value);
+      this.type = key.toLowerCase();
+      return this;
+    }
+
+    /**
+     * Builds a {@link MetadataEntity} from the builder.
+     *
+     * @return {@link MetadataEntity} from the builder
+     * @throws IllegalArgumentException if the key is a CDAP entity and the MetadataEntity is not correct to represent
+     * the CDAP entity
+     */
+    public MetadataEntity build() {
+      if (parts.isEmpty() || type == null || type.isEmpty()) {
+        throw new IllegalArgumentException("key-value pair and type must be specified");
+      }
+      validateHierarchy();
+      return new MetadataEntity(Collections.unmodifiableMap(parts), type);
```

Comment:
you either need to (defensive) copy this map, or make sure that append cannot be called again. otherwise the unmodifiable map can actually be modified.

Reply:
Fixed.